### PR TITLE
feat: ERC677 to native currency

### PR DESCRIPTION
**Pull request by sunguru98**
This is the second part of the ERC677 - Native currency transferability request. I have followed as per the documentation recommended sending back native currency over BSC/Kowan/Mainnet. The UI now has a small toggle over the receiving end, while selecting WKETH/WETH/WBNB over Sokol/xDAI.

When Disabled, it shows the usual WKETH/WETH/WBNB counterpart
Screenshot 2021-04-12 at 12 06 43 PM

When enabled, it shows the native currency option.
Screenshot 2021-04-12 at 12 07 21 PM

Screenshots and Transaction URLs have been posted for the following case. 

1.  WKETH (Sokol) to native currency (ETH Kovan)

- AMB Transaction URL: https://alm-test-amb.herokuapp.com/77/0x1f033af7b25f9565c1f2b70760b9296834fdf527dbdb18a74815f6c29a98de12

- Sokol side transaction: https://blockscout.com/poa/sokol/tx/0x1f033af7b25f9565c1f2b70760b9296834fdf527dbdb18a74815f6c29a98de12/token-transfers

Screenshot 2021-04-12 at 12 04 35 PM

- Kovan side transaction: https://kovan.etherscan.io/tx/0x648c50223ae3b3601ad2fb519012cf3a0903a73c95f599843f028ffb8d6ba754
Screenshot 2021-04-12 at 12 05 01 PM

2. WKETH (Sokol) to ERC677 (WKETH Kovan)

- AMB Transaction URL: https://alm-test-amb.herokuapp.com/77/0x634a4a1aa486faf90feae4e0e56734f5afd26b749d3e20d385012cfedee3d7a6

- Sokol Side transaction: https://blockscout.com/poa/sokol/tx/0x634a4a1aa486faf90feae4e0e56734f5afd26b749d3e20d385012cfedee3d7a6/internal-transactions
Screenshot 2021-04-12 at 12 08 15 PM

- Kovan side transaction: https://kovan.etherscan.io/tx/0x88254ab5e2510d3456e47d654c319b2162b52f5dfc3ca5e0d75ba5d5db68e7d8
Screenshot 2021-04-12 at 12 08 44 PM

The logic for transfer and call has been made dynamic such that the function invokes as per the toToken's symbol .. @dan13ram .. 
Please do lemme know if I sucked anywhere haha.

### CODE REVIEW

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 1]
```diff
@@ -102,7 +132,9 @@ export const ToToken = () => {
               <Text fontSize="lg" fontWeight="bold">
                 {token.name}
               </Text>
-              <AddToMetamask token={token} ml="0.5rem" asModal />
+              {!NATIVE_CURRENCY_SYBMOLS.includes(token.symbol) && (
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:135]: Need to remove symbol checks. They will be triggered for other tokens as well.

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 2]
```diff
@@ -154,6 +188,19 @@ export const ToToken = () => {
                 {utils.formatUnits(amount, token.decimals)}
               </Text>
             )}
+            {bridgeDirection !== ETH_XDAI_BRIDGE && // Temporary block and will be removed once the audit is finished.
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:191]: Should disable from home to foreign and foreign to home for now until audit is finished.
So would suggest keeping this in `lib/networks` and reading from there. 
Something like `isNativeCurrencyBridgingEnabled` ?

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 3]
```diff
@@ -154,6 +188,19 @@ export const ToToken = () => {
...
+              WRAPPED_CURRENCY_SYMBOLS.includes(fromToken.symbol) &&
+              NATIVE_CURRENCY_CHAIN_IDS.includes(chainId) && (
+                <Flex>
+                  <Text>Receive Native Currency</Text>
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:195]: Consider replace `Native Currency` with the actual currency symbol. Such as `Receive KETH` or `Receive BNB` as per the bridge direction.

File: packages/react-app/src/lib/bridge.js [hunk 4]
```diff
@@ -327,14 +328,27 @@ export const fetchTokenLimits = async (
...
+  ethersProvider,
+  token,
+  receiver,
+  amount,
+  shouldSendNativeCurrency,
+) => {
   const signer = ethersProvider.getSigner();
-  const { mode, mediator, address } = token;
+  const { mode, mediator, address, symbol } = token;
+  console.log(token);
```
Comment (dan13ram) [inline packages/react-app/src/lib/bridge.js:340]: unwanted console log. Consider running `yarn lint --fix && yarn format` before pushing.

File: packages/react-app/src/lib/constants.js [hunk 5]
```diff
@@ -19,6 +19,7 @@ export const DEFAULT_BRIDGE_DIRECTION =
   process.env.REACT_APP_DEFAULT_BRIDGE_DIRECTION || ETH_XDAI_BRIDGE;
 
 export const NATIVE_CURRENCY_SYBMOLS = ['ETH', 'KETH', 'BNB'];
+export const WRAPPED_CURRENCY_SYMBOLS = ['WETH', 'WKETH', 'WBNB'];
```
Comment (dan13ram) [inline packages/react-app/src/lib/constants.js:22]: Should not do symbol checks for tokens ever, instead do an address check.

**Comment by sunguru98** [comment]
Pushed the PR review changes for both parent branch as well as this branch ..

File: packages/react-app/src/components/modals/ConfirmTransferModal.jsx [hunk 6]
```diff
@@ -49,9 +51,12 @@ export const ConfirmTransferModal = ({ isOpen, onClose }) => {
   const fromAmt = formatValue(fromAmount, fromToken.decimals);
   const fromUnit = fromToken.symbol;
   const toAmt = formatValue(toAmount, toToken.decimals);
-  const toUnit = NATIVE_CURRENCY_SYBMOLS.includes(toToken.symbol)
-    ? toToken.destinationTokenSymbol
-    : toToken.symbol;
+  const toUnit =
+    providerChainId === foreignChainId &&
+    NATIVE_CURRENCY_SYMBOLS.includes(toToken.symbol)
```
Comment (dan13ram) [inline packages/react-app/src/components/modals/ConfirmTransferModal.jsx:56]: @sunguru98 this also needs to be changed.

File: packages/react-app/src/lib/bridge.js [hunk 7]
```diff
@@ -327,14 +328,33 @@ export const fetchTokenLimits = async (
...
+      ];
+      const helperContract = new Contract(helperContractAddress, abi, signer);
+      return helperContract.wrapAndRelayTokens(receiver, { value: amount });
+    }
     case 'erc677': {
       const abi = ['function transferAndCall(address, uint256, bytes)'];
       const tokenContract = new Contract(address, abi, signer);
-      return tokenContract.transferAndCall(mediator, amount, receiver);
+      const bytesData = shouldSendNativeCurrency
+        ? `${getHelperContract(symbol.replace('W', ''))}${receiver.replace(
```
Comment (dan13ram) [inline packages/react-app/src/lib/bridge.js:352]: Please do not use symbol to get the helper contract, Instead use chainId.

File: packages/react-app/src/lib/constants.js [hunk 8]
```diff
@@ -53,15 +55,16 @@ export const nativeCurrencies = {
...
     destinationTokenAddress: '0xCa8d20f3e0144a72C6B5d576e9Bd3Fd8557E2B04'.toLowerCase(),
     destinationTokenSymbol: 'WBNB',
   },
 };
 
 export const nativeCurrencyMediators = {
-  [NATIVE_CURRENCY_SYBMOLS[0]]: '',
-  [NATIVE_CURRENCY_SYBMOLS[1]]: '0x227a6f13aa0dba8912d740c0f88fb1304b2597e1'.toLowerCase(),
-  [NATIVE_CURRENCY_SYBMOLS[2]]: '0xefc33f8b2c4d51005585962be7ea20518ea9fd0d'.toLowerCase(),
+  [NATIVE_CURRENCY_SYMBOLS[0]]: '',
```
Comment (dan13ram) [inline packages/react-app/src/lib/constants.js:65]: Replace this with chainId as key.

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 9]
```diff
@@ -1,31 +1,70 @@
...
+    () =>
+      getNativeCurrency(
+        NATIVE_CURRENCY_CHAIN_IDS.includes(providerChainId)
+          ? providerChainId
+          : chainId,
+      ),
+    [chainId, providerChainId],
+  );
+
+  const changeToToken = async () => {
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:57]: Could we wrap this inside a useCallback?

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 10]
```diff
@@ -1,31 +1,70 @@
...
   } = useContext(BridgeContext);
   const chainId = getBridgeChainId(providerChainId);
 
   const smallScreen = useBreakpointValue({ base: true, lg: false });
   const [balanceLoading, setBalanceLoading] = useState(false);
 
+  const nativeCurrency = useMemo(
+    () =>
+      getNativeCurrency(
+        NATIVE_CURRENCY_CHAIN_IDS.includes(providerChainId)
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:50]: Can replace this logic with `homeChainId`  from useBridgeDirection.

File: packages/react-app/src/components/bridge/ToToken.jsx [hunk 11]
```diff
@@ -102,7 +141,9 @@ export const ToToken = () => {
               <Text fontSize="lg" fontWeight="bold">
                 {token.name}
               </Text>
-              <AddToMetamask token={token} ml="0.5rem" asModal />
+              {token.address !== ADDRESS_ZERO && (
```
Comment (dan13ram) [inline packages/react-app/src/components/bridge/ToToken.jsx:144]: This check shouldn't be necessary cause AddToMetamask has the check inside the component. It returns `null` if token.address is ADDRESS_ZERO

File: packages/react-app/src/components/modals/ConfirmTransferModal.jsx (hunk 6 above)
Comment (dan13ram) [reply packages/react-app/src/components/modals/ConfirmTransferModal.jsx:56]: `toToken.symbol` itself should be enough right ?

File: packages/react-app/src/lib/helpers.js [hunk 12]
```diff
@@ -160,6 +170,9 @@ export const getRPCKeys = bridgeDirection => {
   }
 };
 
+export const getHelperContract = currencySymbol =>
```
Comment (dan13ram) [inline packages/react-app/src/lib/helpers.js:173]: Lets use chainId as key.

File: packages/react-app/src/lib/token.js [hunk 13]
```diff
@@ -152,6 +151,9 @@ export const fetchTokenBalanceWithProvider = async (
   { address },
   account,
 ) => {
+  if (address === ADDRESS_ZERO) {
```
Comment (dan13ram) [inline packages/react-app/src/lib/token.js:154]: Best if we don't have this logic inside this method. 
This is because there might cases when toToken.address is ADDRESS_ZERO. This is when the fromToken has not yet been bridged onto the other side. In that case toToken doesn't exist yet. So address is ADDRESS_ZERO.

Comment (dan13ram) [reply packages/react-app/src/lib/token.js:154]: So in such cases we need to display the toToken balance as 0.

File: packages/react-app/src/lib/constants.js [hunk 14]
```diff
@@ -14,10 +18,55 @@ export const POLLING_INTERVAL =
 export const DEFAULT_BRIDGE_DIRECTION =
   process.env.REACT_APP_DEFAULT_BRIDGE_DIRECTION || ETH_XDAI_BRIDGE;
 
+export const NATIVE_CURRENCY_SYMBOLS = ['ETH', 'KETH', 'BNB'];
```
Comment (dan13ram) [inline packages/react-app/src/lib/constants.js:21]: We do not need either of these arrays imho. We can use the chainId as key instead of NATIVE_CURRENCY_SYMBOLS, and use foreignChainId instead of NATIVE_CURRENCY_CHAIN_IDS

File: packages/react-app/src/lib/bridge.js [hunk 15]
```diff
@@ -72,6 +81,24 @@ const fetchToTokenDetails = async (bridgeDirection, fromToken, toChainId) => {
     toChainId,
   );
 
+  if (NATIVE_CURRENCY_SYMBOLS.includes(fromSybmol)) {
```
Comment (dan13ram) [inline packages/react-app/src/lib/bridge.js:84]: Please replace this usage to not use symbol

File: packages/react-app/src/components/modals/TokenSelectorModal.jsx [hunk 16]
```diff
@@ -82,11 +112,22 @@ export const TokenSelectorModal = ({ isOpen, onClose, onCustom }) => {
           getGraphEndpoint(chainId),
           getGraphEndpoint(getBridgeChainId(chainId)),
         );
-        const customTokenList = uniqueTokens(
-          baseTokenList.concat(
-            customTokens.filter(token => token.chainId === chainId),
+
+        const nativeCurrency =
+          enableForeignCurrencyBridge &&
+          NATIVE_CURRENCY_CHAIN_IDS.includes(chainId)
```
Comment (dan13ram) [inline packages/react-app/src/components/modals/TokenSelectorModal.jsx:118]:
```suggestion
          foreignChainId === chainId
```

File: packages/react-app/src/components/modals/TokenSelectorModal.jsx [hunk 17]
```diff
@@ -49,26 +59,46 @@ export const TokenSelectorModal = ({ isOpen, onClose, onCustom }) => {
...
         tList.map(async token => ({
           ...token,
-          balance: await fetchTokenBalanceWithProvider(
-            ethersProvider,
-            token,
-            account,
-          ),
+          balance:
+            token.mode === 'NATIVE'
+              ? await ethersProvider.getBalance(account)
```
Comment (dan13ram) [inline packages/react-app/src/components/modals/TokenSelectorModal.jsx:79]: @sunguru98 if we are adding that `mode === NATIVE` check inside `fetchTokenBalanceWithProvider`, then we can revert these changes to directly call `fetchTokenBalanceWithProvider` for all tokens?

File: packages/react-app/src/components/modals/TokenSelectorModal.jsx [hunk 18]
```diff
@@ -49,26 +59,46 @@ export const TokenSelectorModal = ({ isOpen, onClose, onCustom }) => {
...
+                  account,
+                ),
         })),
       );
 
-      return tokenListWithBalance.sort(
-        ({ balance: balanceA }, { balance: balanceB }) =>
-          balanceB.sub(balanceA).gt(0) ? 1 : -1,
+      const natCurIndex = tokenListWithBalance.findIndex(
+        ({ address }) => address === ADDRESS_ZERO,
```
Comment (dan13ram) [inline packages/react-app/src/components/modals/TokenSelectorModal.jsx:89]:
```suggestion
        ({ address }) => address === ADDRESS_ZERO && token.mode === 'NATIVE',
```

File: packages/react-app/src/lib/token.js (hunk 13 above)
Comment (dan13ram) [reply packages/react-app/src/lib/token.js:154]:
```suggestion
  if (address === ADDRESS_ZERO && mode === 'NATIVE') {
```

As you mentioned we can do a mode check as well here.

**Review by dan13ram — APPROVED** [review]
LGTM
